api-server/src/server/component-passport.js

```diff
@@ -118,10 +118,29 @@ export const createPassportCallbackAuthenticator =
           return next(err);
         }
 
-        if (!user || !userInfo) {
+        if (!user && !userInfo) {
           return res.redirect('/signin');
         }
 
+        const state = req && req.query && req.query.state;
+        // returnTo, origin and pathPrefix are audited by getReturnTo
+        let { returnTo, origin, pathPrefix } = getReturnTo(state, jwtSecret);
+        const redirectBase = getPrefixedLandingPath(origin, pathPrefix);
+
+        if (!user && userInfo == 'access_denied') {
```

Comment:
And both query params are available at all times? Together? 

We are only trying to catch the 'access_denied' state. 

This PR only aims at catching one specific use case (Okta mandate) and handles nothing else. 

We currently (current prod) send the user a `/signin` loop for all errors we do not recognize. That should not change with this PR; whether it should at all is a different discussion away from this PR.

Reply:
> I don't think we need to check either...

In that, I should probably just check if there is the `code` param in the query. IDK I feel it's more brittle. No?

Reply:
> And both query params are available at all times? Together?

According to their docs, yes.

> We currently (current prod) send the user a /signin loop for all errors we do not recognize. That should not change with this PR; whether it should at all is a different discussion away from this PR.

Agreed, it's out of scope.

> In that case, I should probably just check if the code param exists in the query. IDK I feel it's more brittle. No?

Nah, we don't care about that.  Also, I believe that's dealt with by passport-auth0.  When we get to this callback we (on successful login) will already have the `accessToken`.

I'll push a couple of commits in a sec and we can discuss more.

Reply:
> I'll push a couple of commits in a sec and we can discuss more.

I'd appreciate it if you test it locally before you push. I more than confident the params are NOT available at all times. Your last two commits will throw.

Reply:
Here is an example:

`/auth/auth0/callback?code=scfe...&state=eyJh...`

Truncating bits to avoid exposing the strings.

Reply:
There's nothing that can throw, but I'll double-check it.

Reply:
Are you referring to the fact that the `error` and `error_description` params won't be present on successful logins?  That's true and expected.  It proceeds to checking if the `user` and `userInfo` exist, which is exactly what we want.

I've verified that the normal login flow works, too.

Reply:
> It proceeds to checking if the user and userInfo exist, which is exactly what we want.

Makes us rely on two sets of two params to be mutually exclusive for the logic to work. That four params to watch for.

And not to mention, it all depends on the packages "looback-component-passport", "passport-autho" and this custom handler to not break things in the future. They have in the past - that's why we have one of them vendored.

I would prefer the logic to be dumb and straight for me. 

So unless checking the `user` and `userInfo` objects was going to introduce a bug, I am not sure why you are refactoring it.

To put it bluntly, this file is my worst nightmare in this codebase; I would like to have it a certain way for my sanity - unless it's introducing a bug, of course.

So do you see a bug in the original logic I have?

Reply:
> And not to mention, it all depends on the packages "looback-component-passport", "passport-autho" and this custom handler to not break things in the future. They have in the past - that's why we have one of them vendored.

Yeah, that's why I've been so slow to respond.  I've spent the entire day digging through the docs and source code.

> So unless checking the user and userInfo objects was going to introduce a bug, I am not sure why you are refactoring it.

That's why I did it.  

The point is `userInfo` is really the `authInfo` provided by loopback-component-passport.  For our purposes it contains `accessToken` (also `identity`, which we ignore). It does not contain the `error` or `error_details`.

Those two, if they're added at all, are added as query parameters to the callback URL, per the docs I linked to earlier.  That's why I changed the logic to check for those query params, rather than user.

If you want other errors (non `blockedByLaw` ones) to redirect to /signin, let me know.

Reply:
> That's why I changed the logic to check for those query params, rather than user.

And wasn't my original code still checking if those parameters exist - for a certain usecase? I am asking only two things:

1. The destructuring doesn't seem to be valid JS syntax because the objects can be undefined, and all of your current code doesn't check before using it.
2. How is the code I wrote buggy? Specifically [freeCodeCamp/freeCodeCamp@`78dede5` (#47992)](https://github.com/freeCodeCamp/freeCodeCamp/pull/47992/commits/78dede58d0f3bf1bc40a46183e54cab11e8c96ed)

Reply:
> `userInfo` will never be `'access_denied'`, so that code will never be executed. It's an object, not a string

That is not true or correct. That's why I have been asking you why for a while. I can see why you may have assumed so.

`userInfo` is a string with the value of `access_denied` when we enable the Auth0 add-on. I have verified this for at least 2 days of testing.

Reply:
So with that sorted, is there any other way you think the logic is buggy?

Reply:
> userInfo is a string with the value of "access_denied" when we enable the Auth0 add-on. I have verified this for at least 2 days of testing.

Where is this documented?  I can't see how this would happen without code changes.

Reply:
It is not to the best I know.

I discovered it - by testing it really close to a production setup. 

And the reason it is this way is the quirky behavior of the package vendoring stuff we have done.

Reply:
Nevermind, I've figured it out.